component/discovery/kubernetes_crds/kubernetes_crds.go

```diff
@@ -0,0 +1,336 @@
+	}
+	c.apply()
+}
+
+func (c *crdManager) addProbe(p *v1.Probe) {
+	probe := c.cg.generateProbeConfig(p)
+	c.mut.Lock()
+	c.discoveryConfigs[probe.JobName] = probe.ServiceDiscoveryConfigs
+	c.scrapeConfigs[probe.JobName] = probe
+	c.mut.Unlock()
+	c.apply()
+}
+
+func (c *crdManager) onAddPodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	level.Info(c.logger).Log("msg", "found pod monitor", "name", pm.Name)
+	c.addPodMonitor(pm)
+}
+
+func (c *crdManager) onUpdatePodMonitor(oldObj, newObj interface{}) {
+	pm := oldObj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.addPodMonitor(newObj.(*v1.PodMonitor))
+}
+func (c *crdManager) onDeletePodMonitor(obj interface{}) {
+	pm := obj.(*v1.PodMonitor)
+	c.clearConfigs("podMonitor", pm.Namespace, pm.Name)
+	c.apply()
+}
+
+func (c *crdManager) onAddServiceMonitor(obj interface{}) {
+	sm := obj.(*v1.ServiceMonitor)
+	level.Info(c.logger).Log("msg", "found service monitor", "name", sm.Name)
+	c.addServiceMonitor(sm)
+}
+func (c *crdManager) onUpdateServiceMonitor(oldObj, newObj interface{}) {
+	sm := oldObj.(*v1.PodMonitor)
+	c.clearConfigs("serviceMonitor", sm.Namespace, sm.Name)
+	c.addServiceMonitor(newObj.(*v1.ServiceMonitor))
+}
+func (c *crdManager) onDeleteServiceMonitor(obj interface{}) {
+	sm := obj.(*v1.ServiceMonitor)
+	c.clearConfigs("serviceMonitor", sm.Namespace, sm.Name)
+	c.apply()
+}
+
+func (c *crdManager) onAddProbe(obj interface{}) {
+	p := obj.(*v1.Probe)
+	level.Info(c.logger).Log("msg", "found probe", "name", p.Name)
+	c.addProbe(p)
+}
+func (c *crdManager) onUpdateProbe(oldObj, newObj interface{}) {
+	p := oldObj.(*v1.Probe)
+	c.clearConfigs("probe", p.Namespace, p.Name)
+	c.addProbe(newObj.(*v1.Probe))
+}
+func (c *crdManager) onDeleteProbe(obj interface{}) {
+	p := obj.(*v1.Probe)
+	c.clearConfigs("probe", p.Namespace, p.Name)
+	c.apply()
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	// innerCtx gets passed to things we create, so we can restart everything anytime we get an update.
+	// Ideally, this component has very little dynamic config, and won't have frequent updates.
+	var innerCtx context.Context
+	// cancel is the func we use to trigger a stop to all downstream processors we create
+	var cancel func()
+	for {
+		select {
+		case <-ctx.Done():
+			if cancel != nil {
+				cancel()
+			}
+			return nil
+		case <-c.onUpdate:
+			if cancel != nil {
+				cancel()
+			}
+			innerCtx, cancel = context.WithCancel(ctx)
+			c.mut.Lock()
+			componentCfg := c.config
+			c.mut.Unlock()
+			crdMan := newManager(c.opts, c.opts.Logger, componentCfg)
+			go crdMan.run(innerCtx)
+		}
+	}
+}
+
+func (c *crdManager) run(ctx context.Context) {
+
+	config, err := c.config.restConfig()
+	if err != nil {
+		level.Error(c.logger).Log("msg", "failed to create rest config", "err", err)
+		return
+	}
+	clientset, err := kubernetes.NewForConfig(config)
+
+	promClientset := versioned.New(clientset.RESTClient())
```

Comment:
This seems risky; if NewForConfig fails, won't clientset be nil? We should move the error check to happen earlier.